### ISSUE TITLE
randomisation py / r

### ISSUE THREAD

**Issue by jwijffels** [mlverse/torch#636]
Is there a way to set seeds the same in R and Python in order that randomisation numbers are the same for testing purposes?

In Python:

```
>>> torch.manual_seed(2019)
<torch._C.Generator object at 0x000000A4FA7B8970>
>>> torch.randn(5)
tensor([-0.1187,  0.2110,  0.7463, -0.6136, -0.1186])
```

In R:

```
> torch_manual_seed(args$seed)
> torch_randn(5)
torch_tensor
-0.3642
-1.1300
-1.9338
 0.5596
 1.1085
[ CPUFloatType{5} ]
```

**Comment by dfalbel** [comment]
Currently there's no way, but that should be possible and we could also make it the default behavior actually. I'll investigate!

**Comment by jwijffels** [comment]
Would be great in order to make sure all model parameters are the same at initialisation at the Python as well as R side when testing / debugging R code which is direct 'copy-pasting' of Python variants.

**Comment by dfalbel** [comment]
Yeah, sure! 

I found the culprit: https://github.com/mlverse/torch/blob/master/R/generator.R#L71-L73
If you comment out those lines you get the same behavior as python.

It should be possible to use the c++ default generator instead of keeping our own in here.

**Comment by dfalbel** [comment]
Yeah, there's a way to query the default generator with `auto gen = at::detail::getDefaultCPUGenerator();` and using `torch_manual_seed()` is sufficient to set its seed.

Instead of maintaining our own generator we could return the default generator here: https://github.com/mlverse/torch/blob/master/src/torch_types.cpp#L830-L836

I'll work on a PR later today.

**Comment by jwijffels** [comment]
Do you mean that will you could solve the initialisation of modules like nn.Linear?

```
>>> torch.manual_seed(2019)
<torch._C.Generator object at 0x0000003262B27950>
>>> torch.randn(3)
tensor([-0.1187,  0.2110,  0.7463])
>>> list(nn.Linear(2, 3, bias=False).named_parameters())
[('weight', Parameter containing:
tensor([[-0.1304, -0.0159],
        [ 0.1245, -0.0746],
        [ 0.3716, -0.6052]], requires_grad=True))]
```

```
> torch:::cpp_torch_manual_seed(as.character(2019))
> torch_randn(3)
torch_tensor
-0.1187
 0.2110
 0.7463
[ CPUFloatType{3} ]
> nn_linear(2, 3, bias = FALSE)$parameters
$weight
torch_tensor
-0.4998  0.0673
 0.3022 -0.6130
 0.5078  0.5416
[ CPUFloatType{3,2} ]
```

**Comment by dfalbel** [comment]
This might solve, if some function there uses the default generator, than it could be using an out of sync one. Although there could be small implementation differences in the `nn_init` methods: https://github.com/mlverse/torch/blob/master/R/nn-init.R that we would need to track down.

**Comment by dfalbel** [comment]
Hey @jwijffels ,

RNG's are now in sync so you should expect to get the same behavior in R and Pytorch. 
I only checked for `nn_linear()` but it seems to also yield the same values.

Please report an issue if you find different behaviors for other modules. Let's consider different initializations given a seed a bug.

**Comment by jwijffels** [comment]
Great. Thanks for the quick interventions!
